### PR TITLE
Register torchvision ops

### LINKED ISSUE #2134: About RegisterOperators in C++
## 🐛 Bug
I'm testing about jit in C++ ( libtorch )

My test cases
```
int main(int argc, const char* argv[]) {
    auto& ops = torch::jit::getAllOperators();
    std::cout << "torch jit operators\n";
    for (auto& op: ops) {
        auto& name = op->schema().name();
        if (name.find("torchvision") != std::string::npos)
            std::cout << "op : " << op->schema().name() << "\n";
    }
    std::cout << "\n";
    return 0;
}

>>> torch jit operators
```

Expected (I changed some codes in vision)
```
>>> torch jit operators
op : torchvision::_cuda_version
op : torchvision::deform_conv2d
op : torchvision::ps_roi_pool
op : torchvision::ps_roi_align
op : torchvision::_new_empty_tensor_op
op : torchvision::roi_pool
op : torchvision::roi_align
op : torchvision::nms
```

In 1st case, seems no jit opeators registry on libtorch.

Is this only occurs my env? or some codes are wrong?

Because I checked expected output when I changed srcs.

## Environment

Install command follows on `reamdme`

torch 1.5.0 and torchvision 0.6.0

### PR DESCRIPTION
**Pull request by bmanga**
This changes the way that the ops are registered with torch. Instead of relying on a static initializer that might not get run based on the linker's mood, we add a RegisterOps function that is automatically run whenever the user includes vision.h exploiting weak symbols.
This way, the user doesn't have to reference any torchvision symbols for the registration to happen.

I have purposely left it flexible enough so that users could just include ops.h and manually call the function with the minimal compile-time overhead.

I had to fix some issues with CMakeLists.txt along the way. Also I have noticed that the functions in the operators headers are not marked inline, which could spell ODR trouble if the users include them themselves, but that's for another PR.

Closes #2134

### CODE REVIEW

**Comment by bmanga** [comment]
I am a bit lost on the python side of things for the CI.
It complains that it can't `#include <torchvision/ops.h>`, but it seems to have no problem with `#include <torchvision/models/models.h>`
From a quick browse around the files I couldn't find why that would be the case.

@fmassa do you happen to know where I should look?

**Comment by fmassa** [comment]
@bmanga `torchvision/ops.h` doesn't exist wrt the include paths that we add in https://github.com/pytorch/vision/blob/3d65fc6723f1e0709916f24d819d6e17a925b394/setup.py#L158
wrt the models, I don't know exactly for sure what's happening, but we don't compile any C++ file for the models in python, so that import is effectively useless wrt Python.

An easy fix would be to `#include "ops.h"` I believe, and maybe also replace `#include <torchvision/models/models.h>` with `#include "models/models.h"`

**Comment by bmanga** [comment]
@fmassa the failures now seem to be unrelated. Does this look ok?

**Comment by fmassa** [comment]
The diff looks good, but let's wait until CI finishes.

The test failures were indeed unrelated, and I've re-run CI to see if it works fine this time (now that the fix to the failing CI issues has been merged in PyTorch)

**Comment by fmassa** [comment]
I have the impression that Windows failures are related, see for example https://app.circleci.com/pipelines/github/pytorch/vision/2855/workflows/cd9d6dff-45a6-45b6-8a03-e716bf82f4da/jobs/159890/steps
```
vision.obj : error LNK2001: unresolved external symbol "public: __cdecl vision::models::VGG13BNImpl::VGG13BNImpl(__int64,bool)" (??0VGG13BNImpl@models@vision@@QEAA@_J_N@Z)
e
```
@peterjc123 do you know what might be the issue? For information, we don't really need the C++ models to be compiled and tested in the python library, so they could be in theory removed.

**Comment by peterjc123** [comment]
Where is `torchvision::_cuda_version()` registered?

**Comment by fmassa** [comment]
@peterjc123 it is currently defined and implemented in `ops.cpp`, maybe we would need to have it live in a header file?

**Comment by peterjc123** [comment]
> @peterjc123 it is currently defined and implemented in `ops.cpp`, maybe we would need to have it live in a header file?

Yep. Or we could also add the annotation `__declspec(dllexport)`.

**Comment by bmanga** [comment]
I forgot that all the code is built by python and not by cmake, so my latest attempt won't work :(

@fmassa Is it possible for the python scripts to rely on cmake in order to avoid maintaining 2 build systems for the c++ side?

**Comment by fmassa** [comment]
@bmanga maybe, but that is a larger endeavor and we will probably not be doing it anytime soon.

**Comment by bmanga** [comment]
@fmassa that's fair. I ended up using C10_EXPORT and C10_IMPORT, but the torchvision_win_py3.6 checks are still failing.
I have no idea what is going wrong. @peterjc123 my last commit is what you had in mind, correct?

I would rather not move the function to a header because users could run into ODR issues if the WITH_CUDA is used for building the lib but not their code or vice versa.

**Comment by fmassa** [comment]
@bmanga about using CMake to compile torchvision, I've opened https://github.com/pytorch/vision/issues/2349 for discussion.

I don't have experience with Windows so I don't know why it's failing. At least, I would expect if the whole function was in the header that this wouldn't be an issue. Also, you mentioned potential issues with `WITH_CUDA`, but note that we already have uses of this macro in headers, so it wouldn't be the first time it happens.
(Note that this might be a bad design choice on my end, and I would be open to alternatives to improve this).

File: torchvision/csrc/vision.cpp [hunk 1]
```diff
@@ -1,55 +1,22 @@
...
 #if PY_MAJOR_VERSION < 3
 PyMODINIT_FUNC init_C(void) {
-  // No need to do anything.
-  // extension.py will run on load
+  // Register torchvision ops for python.
+  vision::RegisterOps();
   return NULL;
 }
 #else
 PyMODINIT_FUNC PyInit__C(void) {
```
Comment (peterjc123) [inline torchvision/csrc/vision.cpp:16]: Those functions will be executed only when you call `import torchvision._C`, not that you load them explicitly through `ctypes.CDLL`. I guess you'll need to move the function call to the global space.

File: torchvision/csrc/ops_autoregister.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+#ifdef _MSC_VER
+#define VISION_INLINE_VARIABLE __declspec(selectany)
+#else
+#define VISION_INLINE_VARIABLE __attribute__((weak))
+#endif
+#endif
+
+namespace vision {
+namespace impl {
+VISION_INLINE_VARIABLE int dummy = RegisterOps();
```
Comment (peterjc123) [inline torchvision/csrc/ops_autoregister.h:15]: So the key problem is why this line is not working, right?

Comment (peterjc123) [reply torchvision/csrc/ops_autoregister.h:15]: I suggest removing the try..catch block in https://github.com/pytorch/vision/blob/master/torchvision/extension.py#L23-L27 to see whether there are skipped errors while loading the extension library.

Comment (peterjc123) [reply torchvision/csrc/ops_autoregister.h:15]: I get an interesting finding here. When I use the Debug configuration, it works. But for Release, it doesn't. So it seems that the variable is optimized away.

File: torchvision/csrc/vision.cpp [hunk 3]
```diff
@@ -1,55 +1,22 @@
...
+#include <Python.h>
 
 // If we are in a Windows environment, we need to define
 // initialization functions for the _C extension
 #ifdef _WIN32
 #if PY_MAJOR_VERSION < 3
 PyMODINIT_FUNC init_C(void) {
-  // No need to do anything.
-  // extension.py will run on load
+  // Register torchvision ops for python.
```
Comment (peterjc123) [inline torchvision/csrc/vision.cpp:11]: Those changes won't work as expected. See https://github.com/pytorch/vision/pull/2253/files#r456852538. Please revert.

File: torchvision/csrc/vision.cpp [hunk 4]
```diff
@@ -1,55 +1,22 @@
...
+  vision::RegisterOps();
   return NULL;
 }
 #else
 PyMODINIT_FUNC PyInit__C(void) {
-  // No need to do anything.
-  // extension.py will run on load
+  // Register torchvision ops for python.
```
Comment (peterjc123) [inline torchvision/csrc/vision.cpp:17]: Those changes won't work as expected. See https://github.com/pytorch/vision/pull/2253/files#r456852538. Please revert.

File: torchvision/csrc/ops_autoregister.h (hunk 2 above)
Comment (peterjc123) [reply torchvision/csrc/ops_autoregister.h:15]: Adding `__declspec(dllexport)` should force the linker not to optimize it away. See https://github.com/pytorch/vision/pull/2489/files#diff-92e9c21d561ea7205d212a64badbc082R13-R17.

**Pull request closed without merging**